Add documentation for restricting job creation without queue name (#2507)
kubernetes-sigs/kueue: site/content/en/docs/tasks/manage/setup_job_admission_policy.md

---
title: "Setup a Job Admission Policy"
date: 2024-06-28
weight: 10
description: >
  Implementing Validating Admission Policy to prevent job creation without queue name.
---

> This page shows you how to set up a [Validating Admission Policy](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy) to prevent the creation of jobs without a queue name within a namespace.

Author: Done.

## Before you begin

Ensure the following conditions are met:

- A Kubernetes cluster is running.

> - The kubectl command-line tool can communicate with your cluster.

Reviewer: What about installing Kueue? :)

## Creating a Validating Admission Policy

> To verify a job contains a queue name, check that the `kueue.x-k8s.io/queue-name` label exists and has a value set using the [Common Expression Language (CEL)](https://github.com/google/cel-spec).

Reviewer 2: The example below shows you how to set up the Job Admission Policy to reject early all Job or JobSets
without the queue-name if sent to the user namespaces. At the same time, it allows you to create and
execute all Jobs in the namespace dedicated to administrative tasks.

In this example we assume that `manageJobsWithoutQueueName` is disabled, letting the admin
to execute Jobs in the `admin-namespace`. Jobs sent to this namespace aren't rejected, or managed
by Kueue.

> without the queue-name if sent to the user namespaces. At the same time, it allows you to create and
> execute all Jobs in the namespace dedicated to administrative tasks.

Reviewer: ```suggestion
without the queue-name if sent to a namespace other than `admin-namespace`.
```

> In this example we assume that `manageJobsWithoutQueueName` is disabled, letting the admin

Reviewer: ```suggestion
You should set `manageJobsWithoutQueueName` to `false` in the Kueue Configuration to let an admin
```

to execute Jobs in the `admin-namespace`. Jobs sent to this namespace aren't rejected, or managed
by Kueue.

{{< include "examples/sample-validating-policy.yaml" "yaml" >}}

To create the policy, download the above file and run the following command:

```shell
kubectl create -f sample-validating-policy.yaml
```

Then, apply the validating admission policy to the namespace by creating a `ValidatingAdmissionPolicyBinding`. The policy binding links the namespaces to the defined admission policy and it instructs Kubernetes how to respond to the validation outcome.

The following is an example of a policy binding:

{{< include "examples/sample-validating-policy-binding.yaml" "yaml" >}}

To create the binding, download the above file and run the following command:

```shell
kubectl create -f sample-validating-policy-binding.yaml
```

> Now, when you try to create a `Job` or a `JobSet` without the `kueue.x-k8s.io/queue-name` label or value in any namespace other than `admin-namespace`, an error message will return:

Reviewer: ```suggestion
Now, when you try to create a `Job` or a `JobSet` without the `kueue.x-k8s.io/queue-name` label or value in any namespace other than `admin-namespace`, the error message will be similar to the following:
```
